akka-remote/src/main/scala/akka/remote/artery/SendQueue.scala

```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.remote.artery
+
+import akka.stream.stage.GraphStage
+import akka.stream.stage.OutHandler
+import akka.stream.Attributes
+import akka.stream.Outlet
+import akka.stream.SourceShape
+import akka.stream.stage.GraphStageLogic
+import org.agrona.concurrent.ManyToOneConcurrentArrayQueue
+import akka.stream.stage.GraphStageWithMaterializedValue
+import org.agrona.concurrent.ManyToOneConcurrentLinkedQueueTail
+import org.agrona.concurrent.ManyToOneConcurrentLinkedQueue
+import java.util.concurrent.atomic.AtomicInteger
+
+/**
+ * INTERNAL API
+ */
+private[remote] object SendQueue {
+  trait OfferApi[T] {
+    def offer(message: T): Boolean
+  }
+
+  private trait WakeupSignal {
+    def wakeup(): Unit
+  }
+}
+
+/**
+ * INTERNAL API
+ */
+private[remote] final class SendQueue[T](capacity: Int) extends GraphStageWithMaterializedValue[SourceShape[T], SendQueue.OfferApi[T]] {
+  import SendQueue._
+
+  val out: Outlet[T] = Outlet("SendQueue.out")
+  override val shape: SourceShape[T] = SourceShape(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, OfferApi[T]) = {
+    @volatile var needWakeup = false
+    val queue = new ManyToOneConcurrentArrayQueue[T](capacity)
+    // TODO perhaps try similar with ManyToOneConcurrentLinkedQueue or AbstractNodeQueue
+
+    val logic = new GraphStageLogic(shape) with OutHandler with WakeupSignal {
+      private val wakeupCallback = getAsyncCallback[Unit] { _ ⇒
+        if (isAvailable(out))
+          tryPush()
+      }
+
+      override def onPull(): Unit =
+        tryPush()
+
+      private def tryPush(): Unit = {
+        needWakeup = true
```

Comment:
this is wasteful, this will cause too many wakeups. You can do the following:
 1. check queue.poll() - it is null
 2. now set needWakeup to true
 3. Do an additional queue.poll() to grab any elements that might missed the needWakeup and have been enqueued just after it.